prometheus_hardware_exporter/collectors/redfish.py

```diff
@@ -67,29 +75,32 @@ def __init__(self, config: Config) -> None:
         self.password = config.redfish_password
         self.timeout = config.redfish_client_timeout
         self.max_retry = config.redfish_client_max_retry
-        self.redfish_obj: HttpClient
-
-    def __enter__(self) -> Self:
-        """Login to redfish while entering context manager."""
         self.redfish_obj = redfish_client(
             base_url=self.host,
             username=self.username,
             password=self.password,
             timeout=self.timeout,
             max_retry=self.max_retry,
         )
-        self.redfish_obj.login(auth="session")
+
+    def __enter__(self) -> Self:
```

Comment:
If I understood correctly you are wrapping redfish_client in this class. As I said on this [comment](https://github.com/canonical/hardware-observer-operator/pull/182#discussion_r1514641149) on `hardware-observer`, I consider this as a bug on the upstream. However, if you are wrapping it, this might be a good opportunity to implement what is necessary to run the disconnect using `__exit__` and pretty much is not necessary to change the `collector.py` file.

Reply:
Yes, I agree, that's another choice we already mentioned and can make, but decided not to introduce `try-except` in `__enter__`, which effectively doing a nested `try-except` block.

```python3
try:
    with RedfishHelper(...) as redfish_helper: <-- another try-except inside or even fix the upstream
....
```

This solution will not introduce less code that the current implementation: adding `try-except` in `__enter__` vs adding `finally` in current `try-except`

(if you exclude my changes to unit test (which I probably should not change, but I did) that remove the context manager)